src/main/java/io/cryostat/diagnostic/DiagnosticsHelper.java

```diff
@@ -165,24 +250,68 @@ public ThreadDump addThreadDump(Target target, String content) {
                 content.length());
     }
 
+    public HeapDump addHeapDump(Target target, FileUpload heapDump, String requestId) {
+        // TODO: Logic to delete the uploaded file after adding to storage
+        // See #1046
+        String filename = heapDump.fileName().strip();
+        if (StringUtils.isBlank(filename)) {
+            throw new BadRequestException();
+        }
+        if (!filename.endsWith(".hprof")) {
+            filename = filename + ".hprof";
+        }
+        log.tracev(
+                "Putting Heap dump into storage with key: {0}", storageKey(target.jvmId, filename));
+        var reqBuilder =
+                PutObjectRequest.builder()
+                        .bucket(heapDumpBucket)
+                        .key(storageKey(target.jvmId, filename))
+                        .contentType(MediaType.TEXT_PLAIN);
+
+        storage.putObject(reqBuilder.build(), RequestBody.fromFile(heapDump.filePath()));
+        var dump =
+                new HeapDump(
+                        target.jvmId,
+                        heapDumpDownloadUrl(target.jvmId, filename),
+                        filename,
+                        clock.now().getEpochSecond(),
+                        heapDump.filePath().toFile().length());
+        bus.publish(
+                MessagingServer.class.getName(),
+                new Notification(
+                        HEAP_DUMP_UPLOADED,
+                        Map.of(
+                                "jobId",
+                                requestId,
+                                "targetId",
+                                target.alias,
```

Comment:
That's misleading - the notification property `targetId` should be `target.id`. There can also be a `targetAlias` property for `target.alias`, or alternatively for connection URL, but ID should mean ID.